cpp/server/server.h

```diff
@@ -0,0 +1,346 @@
+#ifndef CERT_TRANS_SERVER_SERVER_H_
+#define CERT_TRANS_SERVER_SERVER_H_
+
+#include <chrono>
+#include <csignal>
+#include <cstring>
+#include <functional>
+#include <gflags/gflags.h>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <openssl/crypto.h>
+
+#include "config.h"
+#include "log/cert_submission_handler.h"
+#include "log/ct_extensions.h"
+#include "log/file_db.h"
+#include "log/file_storage.h"
+#include "log/frontend.h"
+#include "log/frontend_signer.h"
+#include "log/leveldb_db.h"
+#include "log/log_lookup.h"
+#include "log/log_signer.h"
+#include "log/sqlite_db.h"
+#include "log/tree_signer.h"
+#include "monitoring/latency.h"
+#include "monitoring/monitoring.h"
+#include "monitoring/registry.h"
+#include "server/json_output.h"
+#include "server/proxy.h"
+#include "util/etcd.h"
+#include "util/periodic_closure.h"
+#include "util/read_key.h"
+#include "util/status.h"
+#include "util/thread_pool.h"
+#include "util/uuid.h"
+
+DEFINE_int32(node_state_refresh_seconds, 10,
+             "How often to refresh the ClusterNodeState entry for this node.");
+DEFINE_int32(watchdog_seconds, 120,
+             "How many seconds without successfully refreshing this node's "
+             "before firing the watchdog timer.");
+DEFINE_bool(watchdog_timeout_is_fatal, true,
+            "Exit if the watchdog timer fires.");
+
+namespace cert_trans {
+
+Gauge<>* latest_local_tree_size_gauge =
+    Gauge<>::New("latest_local_tree_size",
+                 "Size of latest locally generated STH.");
+
+
+template <class Logged>
+class Server {
+ public:
+  struct Options {
+    Options() : port(0), num_http_server_threads(16) {
+    }
+
+    std::string server;
+    uint16_t port;
+
+    std::string etcd_root;
+
+    int num_http_server_threads;
+  };
+
+  static void StaticInit();
+
+  // Doesn't take ownership of anything.
+  Server(const Options& opts, std::shared_ptr<libevent::Base> event_base,
+         Database<Logged>* db, EtcdClient* etcd_client,
+         UrlFetcher* url_fetcher, LogSigner* log_signer,
+         CertChecker* cert_checker);
+  ~Server();
+
+  bool IsMaster() const;
+  MasterElection* election();
+  ConsistentStore<Logged>* consistent_store();
+  ClusterStateController<Logged>* cluster_state_controller();
+  LogLookup<Logged>* log_lookup();
+
+  void Initialise();
+  void Run();
+
+
+ private:
+  const Options& options_;
```

Comment:
Ah, no, this is quite forbidden! Since the ctor takes a const ref, you might have a reference to a temporary object here, in other words, invalid memory. :wink: